### PR TITLE
[Placement Group] Remove API part 2

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Previously, when placement groups were removed, we didn't actually kill workers that are already occupying resources. This PR addresses this problem. 

Logic:
Each task spec now will contain placement group id. This way, we can associate all workers with placement groups. When the cancel request is received, we kill all associated workers.

## Limitation
There are still several limitations. This will require me to implement https://github.com/ray-project/ray/issues/10066. 
- All pending tasks or subsequent tasks will be just infeasible. We should remove them properly from local_queue. I will deprioritize this because we need to reimplement after new scheduler is turned on.
- Each core workers still submit pending tasks. For example, core worker will retry task submission although we kill workers. We should disallow this by creating Placement group handle that subscribes the placement group state (similar to actor handles). 

## Related issue number

https://github.com/ray-project/ray/issues/9793

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/common/task/task_spec.h [hunk 1]
```diff
@@ -189,6 +189,8 @@ class TaskSpecification : public MessageWrapper<rpc::TaskSpec> {
   // Compute a static key that represents the given resource shape.
   static SchedulingClass GetSchedulingClass(const ResourceSet &sched_cls);
 
+  const PlacementGroupID PlacementGroupId() const;
```
Comment (ericl) [inline src/ray/common/task/task_spec.h:192]: Please document this.

File: src/ray/raylet/node_manager.cc [hunk 2]
```diff
@@ -1846,8 +1846,28 @@ void NodeManager::HandleCancelResourceReserve(
...
+  // Kill all workers that are currently associated with the placement group.
+  std::vector<std::shared_ptr<WorkerInterface>> workers_associated_with_pg;
+  for (const auto &worker_it : leased_workers_) {
+    auto &worker = worker_it.second;
+    if (worker->GetPlacementGroupId() == bundle_spec.PlacementGroupId()) {
+      workers_associated_with_pg.push_back(worker);
+    }
+  }
+  for (const auto &worker : workers_associated_with_pg) {
+    RAY_LOG(DEBUG)
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc:1859]: RAY_LOG(INFO)

File: src/ray/raylet/node_manager.cc [hunk 3]
```diff
@@ -1846,8 +1846,28 @@ void NodeManager::HandleCancelResourceReserve(
...
+        << "Placement group cancellation disconnects a worker. Plcaement group id: "
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc:1860]:
```suggestion
        << "Destroying worker since its placement group was removed. Placement group id: "
```

**Review by ericl — APPROVED** [review]

File: python/ray/tests/test_placement_group.py [hunk 4]
```diff
@@ -246,33 +246,46 @@ def is_placement_group_removed():
...
+        time.sleep(50)
+
+    # Schedule a long running task and actor.
+    task_ref = long_running_task.options(placement_group_id=pid).remote()
     a = A.options(placement_group_id=pid).remote()
     assert ray.get(a.f.remote()) == 3
+
     ray.experimental.remove_placement_group(pid)
-    # # Subsequent remove request shouldn't do anything
+    # Subsequent remove request shouldn't do anything
```
Comment (ffbin) [inline python/ray/tests/test_placement_group.py:271]: Subsequent remove request shouldn't do anything.

File: src/ray/raylet/node_manager.cc [hunk 5]
```diff
@@ -1843,11 +1843,31 @@ void NodeManager::HandleCancelResourceReserve(
     rpc::CancelResourceReserveReply *reply, rpc::SendReplyCallback send_reply_callback) {
   RAY_CHECK(!new_scheduler_enabled_) << "Not implemented";
   auto bundle_spec = BundleSpecification(request.bundle_spec());
-  RAY_LOG(DEBUG) << "bundle return resource request " << bundle_spec.BundleId().first
-                 << bundle_spec.BundleId().second;
+  RAY_LOG(INFO) << "bundle return resource request " << bundle_spec.BundleId().first
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:1846]: "Cancelling reserved resources, placement group id = " << bundle_spec.BundleId().first << ", bundle index = " << bundle_spec.BundleId().second  ?

**Review by ffbin — APPROVED** [review]
LGTM

**Pull request merged**